app/src/main/java/com/github/gotify/service/GotifyPushNotification.kt

```diff
@@ -0,0 +1,141 @@
+package com.github.gotify.service
+
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.content.ServiceConnection
+import android.os.*
+import androidx.core.os.bundleOf
+import com.github.gotify.log.Log
+
+/**
+ * THIS CLASS IS USED TO PUSH NOTIFICATIONS TO OTHER APPS
+ * It is called from the thread in WebSocketService
+ */
+
+
+//TODO : delete the notification once delivered
+//TODO: Continue to implement the URl change
+// For the moment, re-registering the client is enough
+
+/**
+ * Function to notify client
+ */
+fun notifyClient(context: Context,app: String, msg: com.github.gotify.client.model.Message){
```

Comment:
It is not a service, only a class to bind to client service and handle messages. Also, I wanted to separate the registration service (GotifyMessengerService) cnd the class used to push the notif. We can do it in another way if you prefer to.

Reply:
Okay, I haven't tried using it with another app yet, maybe then I can better understand the structure.
Here an implementation(untested) of this file, with an IMO better structure, because it is more obvious what happens:
```kotlin
interface WithMessenger {
    @Throws(RemoteException::class)
    fun accept(m: Messenger);
}

fun notifyClient(context: Context, clientPackage: String, exec: WithMessenger) {
    val db = MessagingDatabase(context)
    val service = db.getServiceName(clientPackage)
    if (service.isBlank()) {
        Log.w("No service found for $clientPackage")
        return
    }
    val gHandlerThread = HandlerThread(clientPackage)
    gHandlerThread.start()

    val gConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName,
                                        service: IBinder) {
            val gService = Messenger(service)
            Log.i("Remote service connected")
            try {
                exec.accept(gService)
            } catch(e: RemoteException) {
                Log.e("Something went wrong", e)
            } finally {
                context.unbindService(this)
                gHandlerThread.quit()
            }
        }

        override fun onServiceDisconnected(className: ComponentName) {
            gHandlerThread.quit()
        }
    }

    val intent = Intent()
    intent.component = ComponentName(clientPackage, service)
    context.bindService(intent, gConnection, Context.BIND_AUTO_CREATE)
}
```

Reply:
Done, it 's a way better, thanks